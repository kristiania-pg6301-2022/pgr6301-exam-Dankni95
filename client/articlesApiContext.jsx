import React from "react";
import { fetchJSON } from "./lib/fetchJSON";
import { postJSON } from "./lib/postJSON";

export const ArticlesApiContext = React.createContext({
  async fetchLogin() {
    return await fetchJSON("/api/login");
  },
  async listArticles(query) {
    return await fetchJSON("/api/articles?" + new URLSearchParams(query));
  },
  async createArticle(article) {
    return await postJSON("/api/articles", article);
  },
  async deleteArticle(query) {
    return await fetchJSON("/api/articles?" + new URLSearchParams(query));
  },
  async registerLogin(provider, login) {
    return await postJSON(`/api/login/${provider}`, login);
  },
  async endSession() {
    const res = await fetch("/api/login", { method: "DELETE" });
    if (!res.ok) {
      throw new Error(`Failed to post ${res.status}: ${res.statusText}`);
    }
  },
});
