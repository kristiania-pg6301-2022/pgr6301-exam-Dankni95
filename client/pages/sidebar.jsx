import { Button, Grid } from "@mui/material";
import React from "react";

export default ({ user, socket }) => {
  function handleSubmitQuery(e) {
    socket.send(JSON.stringify({ category: e.target.value }));
  }

  return (
    <Grid container spacing={2}>
      <Grid item xs={2}>
        <Button
          onClick={(e) => {
            handleSubmitQuery(e);
          }}
          value=""
        >
          All
        </Button>
        <Grid item xs={8}>
          <Button
            value="Politics"
            onClick={(e) => {
              handleSubmitQuery(e);
            }}
          >
            Politics
          </Button>
        </Grid>
        <Button
          value="Economics"
          onClick={(e) => {
            handleSubmitQuery(e);
          }}
        >
          Economics
        </Button>
        <Grid item xs={8}>
          <Button
            value="Sport"
            onClick={(e) => {
              handleSubmitQuery(e);
            }}
          >
            Sport
          </Button>
        </Grid>
        <Button
          value="Health"
          onClick={(e) => {
            handleSubmitQuery(e);
          }}
        >
          Health
        </Button>
      </Grid>
    </Grid>
  );
};
