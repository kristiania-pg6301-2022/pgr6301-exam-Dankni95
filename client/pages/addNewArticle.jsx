import React, {useContext, useState} from "react";
import {useNavigate} from "react-router-dom";
import {FormInput, FormTextArea, FormTextOptions} from "../lib/formInput";
import {ArticlesApiContext} from "../articlesApiContext.jsx";

export function AddNewArticle({user}) {
    const {createArticle} = useContext(ArticlesApiContext);
    const [title, setTitle] = useState("");
    const [category, setCategory] = useState("Politics");
    const [content, setContent] = useState("");

    const navigate = useNavigate();

    function handleSubmit(event) {
        event.preventDefault();
        const {name} = user.openid
        createArticle({title, date: new Date().toLocaleDateString(), category, content, name});
        navigate("/");
    }

    return (
        <form onSubmit={handleSubmit}>
            <h1>Add new article</h1>

            <FormInput label={"Title:"} value={title} onChangeValue={setTitle}/>
            <FormTextOptions label={"Category: "} value={category} onChangeValue={setCategory}/>
            <FormTextArea label={"Content:"} value={content} onChangeValue={setContent}/>
            <div>
                <button>Submit</button>
            </div>
        </form>
    );
}
