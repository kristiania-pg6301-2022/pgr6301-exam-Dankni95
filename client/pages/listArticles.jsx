import React, {useContext, useState, useEffect} from "react";
import {ArticlesApiContext} from "../articlesApiContext.jsx";
import {useNavigate} from "react-router-dom";
import {useLoading} from "../useLoading";

export async function handleClick(article, deleteArticle) {
    await deleteArticle(article)
    useNavigate()
}


function Article({article: {title, date, category, content, name}, user}) {
    const {deleteArticle} = useContext(ArticlesApiContext)

    return (
        <div>
            <h6>{category}</h6>
            <h1>{"title" + title}</h1>
            <h6>{date}</h6>
            {user?.google || user?.openid ? <h4>{content}</h4> : "Sign in to read the article"}
            <h6>By: {name}</h6>
            {user.openid && user?.openid?.name === name ?
                <button onClick={(event) => handleClick({del: title}, deleteArticle)}>Edit</button> : ""}
        </div>
    );
}


export function ListArticles({user}) {
    const [category, setCategory] = useState("");
    const [categoryQuery, setCategoryQuery] = useState("");


    function handleSubmitQuery(e) {
        e.preventDefault();
        setCategory(categoryQuery);
    }

    const [articles, setArticles] = useState("");
    useEffect(() => {
        const ws = new WebSocket(window.location.origin.replace(/^http/, "ws") + "/");
        ws.onmessage = (event) => {
            const {name, title, content, date, category} = JSON.parse(event.data);
            setArticles((articles) => [...articles, {name, title, content, date, category}]);
        };
    }, []);




    return (
        <div>
            <h1>Articles</h1>

            <div>
                <form onSubmit={handleSubmitQuery}>
                    <label>
                        Category:
                        <select name={"Category"} value={categoryQuery}
                                onChange={(e) => setCategoryQuery(e.target.value)}
                        >
                            <option value="">All</option>
                            <option value="Politics">Politics</option>
                            <option value="Economics">Economics</option>
                            <option value="Sport">Sport</option>
                            <option value="Health">Health</option>
                        </select>
                        <button>Select</button>
                    </label>
                </form>
            </div>
            {Object.keys(articles).map((keyName, i) => (
                <Article key={articles[keyName].title} article={articles[keyName]} user={user}/>
            ))}
        </div>
    );
}
