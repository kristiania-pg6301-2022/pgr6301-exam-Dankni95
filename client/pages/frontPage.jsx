import {Link, Navigate} from "react-router-dom"
import React from "react"
import {ListArticles} from "./listArticles";
import Sidebar from "./sidebar";
import {Card} from "@mui/material";


export function FrontPage({user}) {

    const ws = new WebSocket(window.location.origin.replace(/^http/, "ws"));


    return (
        <>
            <div className="left-side"><Sidebar  socket={ws}/></div>
            <main  style={{ padding: "1rem 1rem 1rem 1rem", width:"98%" }} className="main"><ListArticles user={user} ws={ws} /></main>
            <footer className="section coral">Â© 2022 Kristiania Avis</footer>
        </>
    )
}
