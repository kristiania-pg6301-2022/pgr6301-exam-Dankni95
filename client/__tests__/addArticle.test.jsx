import {AddNewArticle} from "../pages/addNewArticle";
import React from "react";
import ReactDOM from "react-dom";
import {Simulate} from "react-dom/test-utils";
import {MemoryRouter} from "react-router-dom";
import {ArticlesApiContext} from "../articlesApiContext.jsx";

describe("add article component", () => {
    it("shows article form", () => {
        const element = document.createElement("div");
        ReactDOM.render(
            <MemoryRouter>
                <AddNewArticle/>
            </MemoryRouter>,
            element
        );
        expect(element.innerHTML).toMatchSnapshot();
        expect(
            Array.from(element.querySelectorAll("form label strong")).map(
                (e) => e.innerHTML
            )
        ).toEqual(["Title:", "Content:"]);
    });

    it("adds article on submit", () => {
        const createArticle = jest.fn();
        const title = "Test title";
        const content = "test content"
        const category = "Politics"
        const date = new Date().toLocaleDateString()


        const user = {openid: {name: "Danny"}}


        const element = document.createElement("div");
        ReactDOM.render(
            <ArticlesApiContext.Provider value={{createArticle}}>
                <MemoryRouter>
                    <AddNewArticle user={user}/>
                </MemoryRouter>
            </ArticlesApiContext.Provider>,
            element
        );
        Simulate.change(element.querySelector(".form-input input"), {
            target: {value: title},
        });
        Simulate.change(element.querySelector(".form-text-input textarea"), {
            target: {value: content},
        });

        let name = user.openid.name

        Simulate.submit(element.querySelector("form"));
        expect(createArticle).toBeCalledWith({
            category,
            content,
            date,
            name,
            title
        });
    });
});
