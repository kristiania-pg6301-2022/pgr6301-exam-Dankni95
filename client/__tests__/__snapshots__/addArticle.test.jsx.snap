// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add article component shows article form 1`] = `"<form><h1>Add new article</h1><div class=\\"form-input\\"><label><strong>Title:</strong><input value=\\"\\"></label></div><div class=\\"form-input\\"><div><select><option value=\\"Politics\\">Politics</option><option value=\\"Economics\\">Economics</option><option value=\\"Sport\\">Sport</option><option value=\\"Health\\">Health</option></select></div><strong>Category: </strong> </div><div class=\\"form-text-input\\"><label><strong>Content:</strong> <textarea></textarea></label></div><div><button class=\\"submit\\">Submit</button></div></form>"`;
